src/buildstream/storage/directory.py

```diff
@@ -16,328 +17,499 @@
+                            If no filter callback is specified, all files will be imported.
 
-        Yields:
-          A report of files imported and overwritten.
+        Returns:
+           A :class:`.FileListResult` report of files imported and overwritten.
 
+        Raises:
+           DirectoryError: if any system error occurs.
         """
+        return self._import_files_internal(external_pathspec, filter_callback=filter_callback,)
 
-        raise NotImplementedError()
+    def import_single_file(self, external_pathspec: str) -> FileListResult:
+        """Imports a single file from an external path
 
-    def import_single_file(self, external_pathspec, properties=None):
-        """Imports a single file from an external path"""
-        raise NotImplementedError()
+        Args:
+           external_pathspec: A string containing a pathname.
+           properties: Optional list of strings representing file properties to capture when importing.
 
-    def export_files(self, to_directory, *, can_link=False, can_destroy=False):
-        """Copies everything from this into to_directory.
+        Returns:
+           A :class:`.FileListResult` report of files imported and overwritten.
 
-        Args:
-          to_directory (string): a path outside this directory object
-            where the contents will be copied to.
-          can_link (bool): Whether we can create hard links in to_directory
-            instead of copying. Setting this does not guarantee hard links will be used.
-          can_destroy (bool): Can we destroy the data already in this
-            directory when exporting? If set, this may allow data to be
-            moved rather than copied which will be quicker.
+        Raises:
+           DirectoryError: if any system error occurs.
         """
-
         raise NotImplementedError()
 
-    def export_to_tar(self, tarfile, destination_dir, mtime=BST_ARBITRARY_TIMESTAMP):
-        """ Exports this directory into the given tar file.
+    def export_to_tar(self, tarfile: TarFile, destination_dir: str, mtime: int = BST_ARBITRARY_TIMESTAMP) -> None:
+        """Exports this directory into the given tar file.
 
         Args:
-          tarfile (TarFile): A Python TarFile object to export into.
-          destination_dir (str): The prefix for all filenames inside the archive.
-          mtime (int): mtimes of all files in the archive are set to this.
+          tarfile: A Python TarFile object to export into.
+          destination_dir: The prefix for all filenames inside the archive.
+          mtime: mtimes of all files in the archive are set to this.
+
+        Raises:
+           DirectoryError: if any system error occurs.
         """
         raise NotImplementedError()
 
     # Convenience functions
-    def is_empty(self):
-        """ Return true if this directory has no files, subdirectories or links in it.
+    def is_empty(self) -> bool:
```

Comment:
The more Pythonic way would probably be to implement the inverse as `__bool__` (or `__len__` but that might not be needed).

Reply:
Using `__len__` here since that seems to make sense because the `Directory` already otherwise behaves as a collection of entries.

Also updating the docs note about iterating over entries to mention that directories behave like collections of entries, and that they are falsy when emtpy, etc.